### ISSUE TITLE
[Discussion] Performance issues on MacOS/Windows/Linux

### ISSUE THREAD

**Issue by AlexAlabyev** [KELiON/cerebro#118]
let's discuss all the performance issues here, such as:
- Glitches and freezes
- High RAM and CPU usage 

and other similar issues

**Comment by AlexAlabyev** [comment]
Should Cerebro use the electron's latest release **electron v1.6.0 beta**, referring to the issue [https://github.com/KELiON/cerebro/issues/95](https://github.com/KELiON/cerebro/issues/95)? 

It uses the new version of chrome with memory optimizations, according to this page [https://github.com/electron/electron/releases/tag/v1.6.0](https://github.com/electron/electron/releases/tag/v1.6.0)

Sometimes Cerebro takes up a lot of RAM for a short period of time, sitting in the background, maybe that's the problem with electron itself, because I don't see anything in the code of cerebro that would lead to this kind of leaks.

**Comment by KELiON** [comment]
@AlexAlabyev I would wait for production release of electron. 

And I don't think the issue is just in electron. I didn't have time to check, but I suspect it is something in `basic-apps` (applications plugin for both Linux and Windows), because I've never seen Cerebro using more than 200mb on MacOS

**Comment by AlexAlabyev** [comment]
@KELiON okay, so we know that it only happens on windows, so far. @maximbaz have you ever encountered these kind of peaks of RAM and CPU usage on Linux?

**Comment by maximbaz** [comment]
Well it's definitely taking more than 200mb on Linux, most certainly due to the `basic-apps` as you will see below, but I haven't observed any unusual spikes yet. I'll try to pay attention over the next few days. 

I usually see 4 or 5 Cerebro processes, this is the app after maybe a day of normal usage (367mb).
![snapshot](https://cloud.githubusercontent.com/assets/1177900/22900206/c868bfa2-f22d-11e6-8ef8-23c4de8bbe69.png)

Then I close the app, ensure that no processes are left, restart the app, ensure the main window opens (but not doing any searches yet). The memory usage is 450mb. 
![snapshot-3](https://cloud.githubusercontent.com/assets/1177900/22900270/0c5a4280-f22e-11e6-9ebb-159a0e2bda04.png)

After a while of doing nothing it will drop back to ~360mb, which I think is the usual for me. Let's make some stress tests now!

I searched for every english letter, I searched for `a`, `b`, `c`, ..., `z`. My idea was that this would find many apps on my laptop, and thus load many pictures of them in memory. When I went through all the letters, the memory usage is 574mb.
![snapshot-4](https://cloud.githubusercontent.com/assets/1177900/22900330/45a3f4c8-f22e-11e6-8667-b65399c0cfe1.png)

Then I searched for `gif hello` and `gif bye`, looking through every gif result. That lead me to 950mb.
![snapshot-5](https://cloud.githubusercontent.com/assets/1177900/22900842/12edaa54-f230-11e6-9959-e4df70915f99.png)

Then I wrote this text and did something else for 30 minutes, and I still see the memory usage at 950mb. This means a memory leak, the images are not unloaded from the memory once they have been loaded.

**Comment by AlexAlabyev** [comment]
@maximbaz well, that's interesting. Do you know an effective way of purging the memory after each search?
I wish I knew what is causing the random huge spikes when I'm not using it and it sits in the background, as well. I thought it was some silent exception that would look for something once in a while and then when it fails to do something, it goes into a loop and eventually take up some much memory that electron would kill the process and restart it, since it just goes back to normal 250mb after a few seconds. I'm not sure what's that about. Also, it only uses CPU and RAM, it's not doing anything with the disk when that happens, it seems like.

**Comment by maximbaz** [comment]
Speaking of gif plugin, I think the memory leak could be in the [`memoized` function](https://github.com/KELiON/cerebro-gif/blob/master/src/index.js#L30), where the results once loaded are cached... well, forever. I'll leave that to @KELiON to confirm, if true the fix would be to invalidate cache once user is typing a new query (to load a different pack of gifs) or when user is clearing the search box.

Nothing obvious stands out in the `base-apps`, at least at the moment.

**Comment by KELiON** [comment]
@maximbaz just noticed memory usage of gif plugin. I've checked and the reason not in memoize function, it works only with API jsons, and their size usually ~10-13kb. 

Reason in browser caching for gifs. Giphy server adds `Cache-Control:max-age=86400` header, so all gifs are cached for 24 hours. I'll try to find a way to override browser caching, it doesn't make sense to store all gifs for so long period.

**Comment by maximbaz** [comment]
So I managed to decrease the RAM usage of Cerebro after startup from ~430mb down to ~260mb... in a pretty weird way.

Hey @KELiON and @AlexAlabyev, could you check something on your boxes?

When Cerebro is running, I see a few `cerebro` processes, and two of them are `cerebro --type=renderer`:

![image](https://cloud.githubusercontent.com/assets/1177900/22997139/a48f6644-f3d1-11e6-90e7-710d9f84fd86.png)

Do you also have two of them?

Interesting thing, when I kill the first one, the Cerebro window dies, it will be a white empty box.

However if I kill the second one instead... I cannot see any effect of this. Cerebro still works, I can launch apps, plugins work just fine. So I'm curious, what is this process doing, do you know maybe? Is it only me who has it? It consumes 100mb of the memory used by Cerebro, which is a pretty big chunk.

Second thing, see that `cerebro --type=gpu-process` in my picture? It uses another 100mb. Do you also have that one? I can get rid of that process if I start Cerebro with a flag `cerebro --disable-gpu`. I guess there's a cost of doing this, but it's not clearly visible yet.

So when I get rid of the `gpu-process` and one of the `renderer`, I have 3 processes left which consume ~260mb in total.

**Comment by AlexAlabyev** [comment]
@maximbaz Yes, I have 2 render processes on Windows, as well. There's two of them because a renderer is getting created for each webview. So I suppose cerebro is using one renderer for the main process and another one for a webview. You don’t need to use webviews unless you want to run a remote web page, so I guess cerebro needs the webview for something. It initializing everything on start (the one you kill and see no change), so if that's the only thing it does, maybe it should be moved in the main process, instead of using a dedicated webview. 

As for the GPU process, i have it as well, I think it is used for hardware acceleration, not sure if cerebro even needs it enabled. 

Here's some useful information on electron's webviews https://medium.com/@ccnokes/deep-dive-into-electrons-main-and-renderer-processes-7a9599d5c9e2#.s1jk60a4y

**Comment by maximbaz** [comment]
I found what's not working when a second renderer is killed - file search and preview. Now the question is, does this second process provide something else, besides the file operations? 

I would argue that this feature needs to be extracted in a plugin. I'm not sure I want to give 100mb of my RAM for a file search operation that I ~never~ very rarely use in Cerebro.

**Comment by maximbaz** [comment]
I think I'm going to run Cerebro with `--disable-gpu`, so far I don't observe any negative impact on the app, while I do have an extra 100mb free of memory :)

By the way, Slack by default is running with `--disable-gpu`. Perhaps we should consider making it default as well. I'll let you know if I discover any issues with using this flag.

**Comment by KELiON** [comment]
@maximbaz good catch with `--disable-gpu`, I'll investigate how it impacts an app and if it makes sense to use it by default.

About two renderer processes: it is true, one is main Cerebro window and another is just hidden window that is used for running background operations. It is used for all heavy long-running processes, so the main window works fast and without glitches. At least I can mention that it is used by:

* Applications plugins (we are searching for applications in background and send a message to the main process and repeat this process once in a while)
* Files navigation plugin to get list of files in directory, read file;
* Icons reader on MacOS;
* MacOS contacts plugin;

I'm sure that there is a way to reduce memory usage by background processing, we just have to find a better solution. I see two directions to investigate, but maybe you have other ideas:

1. Just find a way to tell Electron that it is just a lightweight process, background window and it won't be shown, so don't load all the APIs; 
1. Use node's `child_process` instead of Electron window. In this case background processing should be rewritten and tested;

**Comment by maximbaz** [comment]
Thanks for the info, so when I kill the second process on linux, not only file navigation stops working, but newly installed applications aren't being discovered anymore. Good to know.

I agree on the directions to investigate. The second point is a bigger investment, but it may have a bigger potential as well, especially in conjunction with #121. I would be interested to test how lightweight are node's `child_process`'es, if they are really lightweight we could think of ways to allow plugins spawn their own processes instead of hosting a background window for all of them. This could both save us resources (if every plugin like apps/file discovery has it's own 10MB process, we don't have to spend resources to host one big background window for them) and stability (if a given plugin is poorly written and gets killed, only its process dies, while all the others survive).

**Comment by maxjoehnk** [comment]
I haven't looked into the code yet but can't the indexing run in the main process?
The renderer process shouldn't care whether the main process is running some search operation and as long as there are no sync FS calls or stuff like that communication between both processes shouldn't be affected.

When you feel like it should run in a second process you need to use child_process
There is no functionality in electron to create a browserwindow which shall never render because child_process does the same thing.

**Comment by KELiON** [comment]
@maxjoehnk not exactly the same, because `BrowserWindow` supports all web APIs (i.e. `fetch`), when `child_process` supports only node

**Comment by maxjoehnk** [comment]
Yeah but there are libraries emulating e.g. the fetch api in node or you need to rewrite the background process to only use node APIs.

Which browser only APIs are you using?
